persistent-postgresql/test/UpsertWhere.hs

```diff
@@ -0,0 +1,146 @@
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      let newItem = Item "item3" "fresh" Nothing Nothing
+      upsertManyWhere
+        (newItem : items)
+        [copyField ItemDescription]
+        []
+        []
+      dbItems <- map entityVal <$> selectList [] []
+      sort dbItems @== sort (newItem : items)
+    it "updates existing records" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      upsertManyWhere
+        items
+        []
+        [ItemQuantity +=. Just 1]
+        []
+    it "only copies passing values" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      let newItems = map (\i -> i { itemQuantity = Just 0, itemPrice = fmap (*2) (itemPrice i) }) items
+          postUpdate = map (\i -> i { itemPrice = fmap (*2) (itemPrice i) }) items
+      upsertManyWhere
+        newItems
+        [ 
+          copyUnlessEq ItemQuantity (Just 0)
+        , copyField ItemPrice
+        ]
+        []
+        []
+      dbItems <- sort . fmap entityVal <$> selectList [] []
+      dbItems @== sort postUpdate
+    it "inserts without modifying existing records if no updates specified" $ runConnAssert $ do
+      let newItem = Item "item3" "hi friends!" Nothing Nothing
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      upsertManyWhere
+        (newItem : items)
+        []
+        []
+        []
+      dbItems <- sort . fmap entityVal <$> selectList [] []
+      dbItems @== sort (newItem : items)
+    it "inserts without modifying existing records if no updates specified and there's a filter" $
+      runConnAssert $ do
+        let newItem = Item "item3" "hi friends!" Nothing Nothing
+        deleteWhere ([] :: [Filter Item])
+        insertMany_ items
+        upsertManyWhere
+          (newItem : items)
+          []
+          []
+          [ItemDescription ==. "hi friends!"]
+        dbItems <- sort . fmap entityVal <$> selectList [] []
+        dbItems @== sort (newItem : items)
+    it "inserts without modifying existing records if there are updates specified but there's a filter" $
+      runConnAssert $ do
+        let newItem = Item "item3" "hi friends!" Nothing Nothing
+        deleteWhere ([] :: [Filter Item])
+        insertMany_ items
+        upsertManyWhere
+          (newItem : items)
+          []
+          [ItemQuantity +=. Just 1]
+          [ItemDescription ==. "hi friends!"]
+        dbItems <- sort . fmap entityVal <$> selectList [] []
+        dbItems @== sort (newItem : items)
+    it "inserts and modifies existing records if there are updates specified and the filter doesn't apply" $
+      runConnAssert $ do
+        let newItem = Item "item3" "hi friends!" Nothing Nothing
+        deleteWhere ([] :: [Filter Item])
+        insertMany_ items
+        upsertManyWhere
+          (newItem : items)
+          []
+          [ItemQuantity +=. Just 1]
+          [ItemDescription ==. "bye friends!"]
```

Comment:
So, with this in `[Filter Item]`, we won't ever apply the update, because none of the `ItemDescription`s are equal to `"bye friends!"`.

We don't actually have any tests that show that the `[Update Item]` is actually being performed, which I think is a bit of a problem. We do test that `[HandleUpdateCollision Item]` is being performed.

I'd like to see one more test - one that has an item in each of the lists - and then an assertion that each item was successfully used. I can merge this into a temp branch and then finish that out